tao_compiler/mlir/disc/transforms/disc_transform_legalize_to_loop.cc

```diff
@@ -0,0 +1,242 @@
+  return pm.run(m);
+}
+
+LogicalResult DiscTransformLegalizeToLoopPass::inlineTransformedModule(
+    OpBuilder& b, Operation* fusion, FusionPattern& fusionPattern, ModuleOp m) {
+  auto funcOps = llvm::to_vector<4>(m.getOps<func::FuncOp>());
+  if (funcOps.size() != 1)
+    return fusion->emitError() << "failed to inline the transformed module "
+                                  "with multiple functions\n";
+  if (funcOps[0].getBody().getBlocks().size() != 1)
+    return fusion->emitError()
+           << "failed to inline the transformed func with multiple blocks\n";
+
+  Block* body = cast<lmhlo::FusionOp>(fusion).getBody();
+  for (auto& nestedOp : llvm::make_early_inc_range(body->without_terminator()))
+    nestedOp.erase();
+  b.setInsertionPoint(body, body->begin());
+  auto mapping = buildValueMapping(fusionPattern, funcOps[0], true);
+  for (auto& nestedOp : funcOps[0].getBody().front().without_terminator()) {
+    Operation* cloned = b.clone(nestedOp, mapping);
+  }
+  return success();
+}
+
+LogicalResult DiscTransformLegalizeToLoopPass::handleCpuFusionOp(
+    OpBuilder& b, Operation* fusion, ShapeAnalysis& shapeAnalysis) {
+  auto fusionOp = cast<lmhlo::FusionOp>(fusion);
+  assert(fusionOp);
+  FusionPattern fusionPattern(fusionOp, &shapeAnalysis);
+  if (!fusionPattern.isTransformBasedFusion()) {
+    // skip non-transform-based fusion pattern.
+    return success();
+  }
+
+  // 1, Outline the fusion to a standalone module op.
+  OwningOpRef<ModuleOp> m;
+  if (failed(outlineFusionOp(fusionOp, fusionPattern, m))) return failure();
+  LLVM_DEBUG(llvm::dbgs() << "After outline fusion op:\n" << m.get() << "\n");
+
+  // 2, TODO(wyzero): assign a default schedule for each pattern here.
+
+  // 3, Build a nested pass pipeline to legalize the outlined fusion op.
+  if (failed(runTransformPipeline(m.get()))) return failure();
+  LLVM_DEBUG(llvm::dbgs() << "After run transform pipeline:\n"
+                          << m.get() << "\n");
+
+  // 4, Inline the lowered IR into the orignal module.
+  if (failed(inlineTransformedModule(b, fusion, fusionPattern, m.get())))
+    return failure();
+  LLVM_DEBUG(llvm::dbgs() << "After inline transformed module:\n"
+                          << *fusion << "\n");
+  return success();
+}
+
+void DiscTransformLegalizeToLoopPass::runOnOperation() {
+  func::FuncOp func = getOperation();
+  OpBuilder b(func);
+  SmallVector<Operation*, 4> gpu_fusion_worklist;
+  SmallVector<Operation*, 4> cpu_fusion_worklist;
+  func.walk([&](lmhlo::FusionOp op) {
+    if (placement_utils::isGpuLmhlo(op))
+      gpu_fusion_worklist.push_back(op);
+    else
+      cpu_fusion_worklist.push_back(op);
+  });
+
+  std::unique_ptr<ShapeAnalysis> shapeAnalysisPtr;
+  if (!gpu_fusion_worklist.empty() || !cpu_fusion_worklist.empty()) {
+    if (useShapeConstraintIR()) {
+      shapeAnalysisPtr.reset(new ShapeConstraintIRAnalysis(func));
+    }
+  }
+
+  for (Operation* fusion : gpu_fusion_worklist) {
```

Comment:
what if stitch is enabled on CPU ?

Reply:
`kStitch` and `kTransform` are processed by different pass. This pass has nothing to do with `kStitch`.